src/netbase.cpp

```diff
@@ -1072,3 +1072,131 @@ void InterruptSocks5(bool interrupt)
+                "Send interrupted (sent only %u of %u bytes before that)", sent, data.size()));
+        }
+
+        // Wait for a short while (or the socket to become ready for sending) before retrying
+        // if nothing was sent.
+        Sockets sockets;
+        sockets.send = {socket};
+        SocketEvents(sockets, std::min(deadline - now, MAX_WAIT_FOR_IO), interrupt);
+    }
+}
+
+std::string RecvUntilTerminator(const SOCKET& socket,
+                                uint8_t terminator,
+                                const std::chrono::milliseconds& timeout,
+                                CThreadInterrupt& interrupt)
+{
+    const auto deadline = GetTime<std::chrono::milliseconds>() + timeout;
+    std::string data;
+    bool terminator_found = false;
+
+    // We must not consume any bytes past the terminator from the socket.
+    // One option is to read one byte at a time and check if we have read a terminator.
+    // However that is very slow. Instead, we peek at what is in the socket and only read
+    // as much bytes as possible without crossing the terminator.
+    // Reading 64 MiB of random data with 262526 terminator chars takes 37 seconds to read
+    // one byte at a time VS 0.71 seconds with the "peek" solution below. Reading one byte
+    // at a time is about 50 times slower.
+
+    for (;;) {
+        char buf[512];
+
+        const ssize_t peek_ret = recv(socket, buf, sizeof(buf), MSG_PEEK);
+
+        switch (peek_ret) {
+        case -1: {
+            const int err = WSAGetLastError();
+            if (IOErrorIsPermanent(err)) {
+                throw std::runtime_error(strprintf("recv(): %s", NetworkErrorString(err)));
+            }
+            break;
+        }
+        case 0:
+            throw std::runtime_error("Connection unexpectedly closed by peer");
+        default:
+            auto end = buf + peek_ret;
+            auto terminator_pos = std::find(buf, end, terminator);
+            terminator_found = terminator_pos != end;
+
+            const size_t try_len =
+                terminator_found ? terminator_pos - buf + 1 : static_cast<size_t>(peek_ret);
+
+            const ssize_t read_ret = recv(socket, buf, try_len, 0);
+
+            if (static_cast<size_t>(read_ret) != try_len) {
+                throw std::runtime_error(
+                    strprintf("recv() returned %u bytes on attempt to read %u bytes but previous "
+                              "peek claimed %u bytes are available",
+                              read_ret, try_len, peek_ret));
+            }
+
+            // Don't include the terminator in the output.
+            const size_t append_len = terminator_found ? read_ret - 1 : read_ret;
+
+            data.append(buf, buf + append_len);
+
+            if (terminator_found) {
+                return data;
+            }
+        }
+
+        const auto now = GetTime<std::chrono::milliseconds>();
+
+        if (now > deadline) {
```

Comment:
I think that in case it reaches the deadline it should throw because otherwise the timeout will be infringed.
```c++
if (now >= deadline)
```

Reply:
Right, and also `SendComplete()` uses `>=`. Changed.